text/0000-security-advisories.md

```diff
@@ -0,0 +1,259 @@
+dependencies, slows down the propagation of critical security fixes.
+
+Assume a crate `W`, which depends on `X`, which depends on `Y`, which depends
+on `Z`.  If `Z` releases a new MINOR version including a security fix, it
+requires the attention of `Y`'s  and `X`'s maintainers to propagate that
+security fix to `W`. What makes this situation worse is that the author of `W`
+is never notified that they were running a vulnerable version of `Z` all the
+time.
+
+Tooling that builds on top of the API provided by Crates.io (such as Cargo)
+could alert crate users of their vulnerabilities, which in turn spurs them to
+update their dependencies accordingly. Even if that does not happen, the
+additional metadata at least makes it clear which crates are potentially
+dangerous to use and which ones not. This not only helps Rust programmers, but
+potentially also distributors (such as packagers of Linux distros) and
+end-users.
+
+# Detailed design
+[design]: #detailed-design
+
+## Crates.io
+
+Similar to yanking, Crates.io should provide an API that allows a user of Cargo
+to attach an arbitrary amount of so-called security advisories to crates they
+own.
+
+Each advisory gets assigned an ID that is unique within the set of advisories
+for the affected crate.  Every advisory should have a unique URL, for example
+`https://crates.io/crates/<crate>/advisory/<id>`, where `<id>` is the
+advisory's ID. On that URL a human-readable representation of the advisory
+should be stored.
+
+Other pages on Crates.io should link to those advisories prominently where
+appropriate.
+
+## Cargo
+
+### `cargo advisory`
+
+A command called `advisory` will be added to Cargo. Here is an excerpt of its help
+page:
+
+```
+$ cargo advisory --help
+Generate and upload security advisories for the given or the current crate.
+
+Usage:
+    cargo advisory [options] -- [<crate>]
+    --filename PATH      The filename to use. Defaults to `./Advisory.toml`.
+                         If `-` is given, generated advisories are printed to
+                         stdout and advisories to upload are read from stdin.
+
+    --vers VERSION       Versions to release this advisory for. Can be
+                         specified multiple times. Only valid in conjunction
+                         with --generate.
+
+    --upload/--generate  Whether to upload or generate a advisory. The default
+                         is to generate. These options are mutually exclusive.
+    [...]
+```
+
+Like `yank` it takes a `--vers` option, with two differences:
+
+- if a version is not specified, `advisory` will default to all existing
+  versions.
+
+- Version ranges such as `<1.2.6, >1.0.0` can be specified. This is comparable
+  to the syntax used for specifying dependencies in the `Cargo.toml`, with the
+  exception that `x.y.z` is not equivalent to `^x.y.z`, but means the exact
+  version.
+
+Here's the workflow:
+
+1. The user invokes `cargo advisory` without the `--upload` option. Cargo will
```

Comment:
The idea behind that awkward phrasing is that both `cargo advisory` and `cargo advisory --generate` are valid for this step. "without --upload" is a phrasing that captures both.

(damn "start a review" button)